ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueCommitting.scala

```diff
@@ -0,0 +1,279 @@
+          .setRecordTime(buildTimestamp(recordTime))
+          .setRejectionEntry(
+            buildRejectionEntry(txEntry, RejectionReason.DuplicateCommand)
+          )
+          .build,
+        Map.empty
+      )
+
+    // FIXME(JM): Refactor this?
+    (dedupCheckResult, letCheckResult, activenessCheckResult, modelCheckResult) match {
+      case (false, _, _, _) =>
+        reject(RejectionReason.DuplicateCommand)
+
+      case (_, false, _, _) =>
+        reject(RejectionReason.MaximumRecordTimeExceeded)
+
+      case (_, _, false, _) =>
+        reject(RejectionReason.Inconsistent)
+
+      case (true, true, true, Left(err)) =>
+        reject(RejectionReason.Disputed(err.msg)) // FIXME(JM): or detailMsg?
+
+      case (true, true, true, Right(())) =>
+        // All checks passed. Return transaction log entry, and update the DAML state
+        // with the committed command and the created and consumed contracts.
+
+        var stateUpdates = scala.collection.mutable.Map.empty[DamlStateKey, DamlStateValue]
+        stateUpdates += commandDedupKey(txEntry.getSubmitterInfo) -> emptyDamlStateValue
+        val effects = InputsAndEffects.computeEffects(entryId, relTx)
+        effects.consumedContracts.foreach { key =>
+          val cs = inputState(key).fold(DamlContractState.newBuilder)(_.getContractState.toBuilder) // FIXME(JM): Previous contract state should always be there
+          cs.setArchivedAt(buildTimestamp(recordTime))
+          cs.setArchivedByEntry(entryId)
+          stateUpdates += key -> DamlStateValue.newBuilder.setContractState(cs.build).build
+        }
+        effects.createdContracts.foreach { key =>
+          val cs = DamlContractState.newBuilder
+          cs.setActiveAt(buildTimestamp(recordTime))
+          stateUpdates += key -> DamlStateValue.newBuilder.setContractState(cs).build
+        }
+
+        (
+          DamlLogEntry.newBuilder
+            .setRecordTime(buildTimestamp(recordTime))
+            .setTransactionEntry(txEntry)
+            .build,
+          stateUpdates.toMap
+        )
+    }
+  }
+
+  // ------------------------------------------------------
+
+  private def buildRejectionEntry(
+      txEntry: DamlTransactionEntry,
+      reason: RejectionReason): DamlRejectionEntry = {
+    val builder = DamlRejectionEntry.newBuilder
+    builder
+      .setSubmitterInfo(txEntry.getSubmitterInfo)
+
+    reason match {
+      case RejectionReason.Inconsistent =>
+        builder.setInconsistent("")
+      case RejectionReason.Disputed(disputeReason) =>
+        builder.setDisputed(disputeReason)
+      case RejectionReason.ResourcesExhausted =>
+        builder.setResourcesExhausted("")
+      case RejectionReason.MaximumRecordTimeExceeded =>
+        builder.setMaximumRecordTimeExceeded("")
+      case RejectionReason.DuplicateCommand =>
+        builder.setDuplicateCommand("")
+      case RejectionReason.PartyNotKnownOnLedger =>
+        builder.setPartyNotKnownOnLedger("")
```

Comment:
We seem to not be creating this rejection-reason. Should we? Should we add a link to an issue? Or a least a TODO?